source/grid/grid_generator.cc

```diff
@@ -7090,6 +7090,283 @@ namespace GridGenerator
+
+            local_vertex_indices[cell.n_vertices() + f] =
+              face_to_new_vertex_indices[f_global];
+          }
+
+        // (iii) create new midpoint vertex locations for each cell
+        if (dim == 2)
+          {
+            local_vertex_indices[cell.n_vertices() + cell.n_faces()] =
+              vertices.size();
+            vertices.push_back(cell.center(/*respect_manifold*/ true));
+          }
+
+        // helper function for creating cells and subcells
+        const auto add_cell =
+          [&](const unsigned int               struct_dim,
+              const std::vector<unsigned int> &index_vertices,
+              const unsigned int               material_or_boundary_id) {
+            if (struct_dim == dim) // cells
+              {
+                CellData<dim> cell_data(index_vertices.size());
+                for (unsigned int i = 0; i < index_vertices.size(); ++i)
+                  {
+                    cell_data.vertices[i] =
+                      local_vertex_indices[index_vertices[i]];
+                    cell_data.material_id = material_or_boundary_id;
+                  }
+                cells.push_back(cell_data);
+              }
+            else if (dim == 2 && struct_dim == 1) // an edge of a triangle
+              {
+                CellData<1> boundary_line(2);
+                boundary_line.boundary_id = material_or_boundary_id;
+                for (unsigned int i = 0; i < index_vertices.size(); ++i)
+                  {
+                    boundary_line.vertices[i] =
+                      local_vertex_indices[index_vertices[i]];
+                  }
+                subcell_data.boundary_lines.push_back(boundary_line);
+              }
+            else if (dim == 3 && struct_dim == 2) // a face of a thetrahedron
+              {
+                CellData<2> boundary_quad(3);
+                boundary_quad.boundary_id = material_or_boundary_id;
+                for (unsigned int i = 0; i < index_vertices.size(); ++i)
+                  {
+                    boundary_quad.vertices[i] =
+                      local_vertex_indices[index_vertices[i]];
+                  }
+                subcell_data.boundary_quads.push_back(boundary_quad);
+              }
+            else
+              {
+                Assert(false, ExcNotImplemented());
+              }
+          };
+
+        const auto current_material_id = cell.material_id();
+
+        // create cells one by one
+        if (dim == 2)
+          {
+            add_cell(dim, {0, 6, 4}, current_material_id);
+            add_cell(dim, {8, 4, 6}, current_material_id);
+            add_cell(dim, {8, 6, 5}, current_material_id);
+            add_cell(dim, {1, 5, 6}, current_material_id);
+            add_cell(dim, {2, 4, 7}, current_material_id);
+            add_cell(dim, {8, 7, 4}, current_material_id);
+            add_cell(dim, {8, 5, 7}, current_material_id);
+            add_cell(dim, {3, 7, 5}, current_material_id);
```

Comment:
I like @bangerth' idea: https://github.com/dealii/dealii/pull/11201/commits/de3ee8db347a517de5ef5995d114fa5bc2c45444#r532703430

Could you convert these types of codes to:
```cpp
            std::array<std::array<unsigned int, 8>, 3> table = {{0,6,4},{8,4,6}, ...};
            for(const auto & cell : table)
              add_cell(dim, table, current_material_id);
```
and in similar the face definitions.

Reply:
I assume @peterrum meant
```
            constexpr std::array<std::array<unsigned int, 8>, 3> table = {{0,6,4},{8,4,6}, ...};
            for(const auto & cell : table)
              add_cell(dim, cell, current_material_id);
```